docs/architecture/adr-059-go-api-stability.md

```diff
@@ -0,0 +1,193 @@
+
+* **External package:** a different Go package, can be a child or sibling package in the same project.
+
+* **Internal code:** code not intended for use in external projects.
+
+* **Internal directory:** code under `internal/` which cannot be imported in external projects.
+
+* **Exported:** a Go identifier starting with an uppercase letter, which can therefore be accessed by an external package.
+
+* **Private:** a Go identifier starting with a lowercase letter, which therefore cannot be accessed by an external package unless via an exported field, variable, or function/method return value.
+
+* **Public API:** any Go identifier that can be imported or accessed by an external project, except test code in `_test.go` files.
+
+* **Private API:** any Go identifier that is not accessible via a public API, including all code in the internal directory.
+
+## Alternative Approaches
+
+- Split all public APIs out to separate Go modules in separate Git repositories, and consider all Tendermint code internal and not subject to API backwards compatibility at all. This was rejected, since it has been attempted by the Tendermint project earlier, resulting in too much dependency management overhead.
+
+- Simply document which APIs are public and which are private. This is the current approach, but users should not be expected to self-enforce this, and external projects will often end up depending on internal code anyway.
+
+## Decision
+
+From Tendermint 1.0, all internal code (except private APIs) will be placed in a root-level [`internal` directory](https://golang.org/cmd/go/#hdr-Internal_Directories), which the Go compiler will block for use by external projects. All exported items outside of the `internal` directory are considered a public API and subject to backwards compatibility guarantees, except files ending in `_test.go`.
+
+The `crypto` package will be split out to a separate module in a separate repo. This is the main general-purpose package used by external projects, and is the only Tendermint dependency in e.g. IAVL which can cause some problems for projects depending on both IAVL and Tendermint.
+
+The `tm-db` package will remain a separate module in a separate repo.
+
+## Detailed Design
+
+### Public API
+
+When preparing our public API for 1.0, we should keep these principles in mind:
+
+- Limit the number of public APIs that we start out with - we can always add new APIs later, but we can't change or remove APIs once they're made public.
+
+- Before an API is made public, do a thorough review of the API to make sure it covers any future needs, can accomodate expected changes, and follows good API design practices.
+
+The following current packages will be part of our public API in 1.0 in some form:
+
+- `abci`
+- `config`
+- `crypto` (separate module)
+- `libs/log`
+- `light`
+- `node`
+- `privval` (client parts)
+- `rpc/client`
+- `types` (respective packages)
+- `version`
+
```

Comment:
In this design how do you imagine developers adding in custom reactors. Will each module interface be housed within node?

If I want to create a custom data incentivization reactor how will I know what interfaces are exposed by the core modules? 

sorry if im asking something that has been stated already.

Reply:
I suggest we postpone public interfaces for e.g. reactors, mempools, and similar until after 1.0. This is something we can add in e.g. 1.1 when we have time to design the API properly, and I'm wary of trying to chew off too much for 1.0.

Reply:
Can we state that the feature to add custom reactors is being removed till a further notice Back when this was added, it was asked for by a couple users (Loom, Corestar).

Reply:
I suggest we discuss this on Monday. If custom reactors is something that many users already rely on, and forking the project is not viable, then we must provide an API for this. However, designing this API properly is a large project that also has relevance for the P2P refactoring, so if we _can_ postpone it until after 1.0 then I think we should.

Reply:
I am fine with postponing it, I agree it is a hard thing to do. I think mentioning it in the ADR could be useful. This could also be added to the negative section as it is removing a feature.

Reply:
Sure, added a paragraph on it. This was already listed as a negative.

Reply:
I think there's a genuine case to be made for allowing Public Go APIs that are not covered by SemVer so that we can still enable this kind of pluggability but not feel locked in, and just isolate and clarify where these situations exist as much as possible.

Reply:
Not sure I agree. SemVer is all about communication: by adhering to it, our users know exactly what to expect. Being selective about where it applies makes that communication ambiguous, confusing our users as well as us about what is and isn't acceptable to change in a given release. We currently use a variant of the approach you outline, where we document which APIs are subject to change, but even the SDK team gets upset when we break APIs that we've clearly documented that we can break.

I also believe Go requires modules to use strict semantic versioning, and rely on this assumption when resolving dependencies. This doesn't affect us much while we're on 0.x, since anything goes, but with v1 and v2 we'll start seeing the effects of this.

As for pluggability, we've had some internal discussions on this which are summarized in this document: https://docs.google.com/document/d/1G08LnwSyb6BAuCVSMF3EKn47CGdhZ5wPZYJQr4-bw58/edit?ts=5f609f11